### PR TITLE
Add isBlockhashValid method to web3.js

### PR DESCRIPTION
**Pull request by zhe-t**
#### Problem
Attempting to resolve #25096 - `Add isBlockhashValid support`

#### Summary of Changes
Added `isBlockhashValid` method to web3.js sdk

Fixes #

### CODE REVIEW

File: web3.js/src/connection.ts [hunk 1]
```diff
@@ -1826,6 +1826,11 @@ const GetLatestBlockhashRpcResult = jsonRpcResultAndContext(
   }),
 );
 
+/**
+ * Expected JSON RPC response for the "isBlockhashValid" message
+ */
+const IsBlockhashValidRpcResult = jsonRpcResult(boolean());
+
```
Comment (steveluscher) [inline web3.js/src/connection.ts:1833]: I know that you're just following what others have done before, but can you _inline_ this into `isBlockhashValid`? It's only ever used once, not shared, and doing work (ie. calling `jsonRpcResult()` and `boolean()` in the module factory) is actually bad and going to be something we change in #20673.

File: web3.js/src/connection.ts [hunk 2]
```diff
@@ -3359,6 +3364,25 @@ export class Connection {
...
   }
 
+  /**
+   * Check whether a blockhash is still valid or not
+   */
+  async isBlockhashValid(
+    blockhash: string,
+    commitment?: Commitment,
+    minContextSlot?: number
+  ): Promise<boolean> {
```
Comment (steveluscher) [inline web3.js/src/connection.ts:3374]: This ‘multiple optional arguments’ signature can be troublesome if you want to supply a `minContextSlot` but you _don't_ want to supply a commitment.

```
const isValid = await isBlockhashValid('ABC', undefined, 1143);
```

What's `undefined`? Should I label it with a comment? Awkward!

Instead, use the `configOrCommitment` pattern (see `getProgramAccounts`).

File: web3.js/src/connection.ts [hunk 3]
```diff
@@ -3359,6 +3364,25 @@ export class Connection {
...
+    const args = this._buildArgs([blockhash], commitment, minContextSlot);
+    const unsafeRes = await this._rpcRequest('isBlockhashValid', args);
+    const res = create(unsafeRes, jsonRpcResult(boolean()));
+    if ('error' in res) {
+      throw new Error(
+        'failed to check if blockhash is valid: ' + res.error.message,
```
Comment (steveluscher) [inline web3.js/src/connection.ts:3380]:
```suggestion
        'Could not determine if blockhash `' + blockhash + '` was valid: ' + res.error.message,
```

**Review by steveluscher — DISMISSED** [review]
Thanks, @zhe-t!

Now all this PR needs is some tests. Use the ‘get block time’ test in `connection.test.ts` as a starting point. Run both `yarn test` and `yarn:test-live-with-test-validator`. You can turn on _only_ the test you're working on by adding `.only` to the end of `it` or `describe`.

Write tests that cover:

* the RPC returning a value when you supply a blockhash
* the RPC returning a value when you supply a blockhash and a min slot
* (LIVE mode only) getting the recent blockhash then asking if it's valid (expect to be true)
* (LIVE mode only) getting the recent blockhash and slot then asking if it's valid with a min slot one slot higher (expect to be false)
* (LIVE mode only) supplying a junk blockhash then asking if it's valid (expect to be false)

File: web3.js/src/connection.ts (hunk 1 above)
Comment (zhe-t) [reply web3.js/src/connection.ts]: No problem, removed and inlined.

File: web3.js/src/connection.ts (hunk 2 above)
Comment (zhe-t) [reply web3.js/src/connection.ts]: Refactored optional arguments using configOrCommitment

File: web3.js/src/connection.ts (hunk 3 above)
Comment (zhe-t) [reply web3.js/src/connection.ts]: Updated error message

**Comment by zhe-t** [comment]
Thanks @steveluscher - working on tests now.

**Comment by zhe-t** [comment]
@steveluscher could you help me with the 
> (LIVE mode only) getting the recent blockhash and slot then asking if it's valid with a min slot one slot higher (expect to be false)
test please? 

The test is called `is blockhash valid - blockhash only and mint slot (live)` in https://github.com/solana-labs/solana/pull/25888/commits/ca50893a0de7671bceaabbe557f0e8b6358da1c8#diff-0ff00ad7672d836c05bc2a1180365eb6a757b5e06c80edc7e9b8a7456b9a52d6R2991. 

I must have over looked something. The recent slot is always 0 when using `yarn:test-live-with-test-validator`.

**Comment by steveluscher** [comment]
> The recent slot is always 0 when using yarn:test-live-with-test-validator.

For sure! In cases where it's the _first_ test you run, this will be the case.

Just loop, waiting for a minimum of slots to pass.

https://github.com/solana-labs/solana/blob/ed5fcde700f6d0dec93c37bde3846082ef3e3abc/web3.js/test/connection.test.ts#L2969-L2971

**Comment by zhe-t** [comment]
I've tried using the slot returned by `getLatestBlockhashAndContext` and using `getSlot` (which are both now non-0) but the method returns a true value when using this slot value+1. Do you have an idea of what I have overlooked? @steveluscher

**Comment by zhe-t** [comment]
cc @steveluscher

**Comment by steveluscher** [comment]
Sorry for the delay here!

I don't really know, but here are some thoughts:

* One of the values that you get back from `getLatestBlockhash` is one called `lastValidBlockheight`. What happens if you set `minContextSlot` to `lastValidBlockheight + 1`?
* Do you have to wait for the `lastValidBlockheight` to _pass_ before making your assertion?

I don't really know what the effect of `minContextSlot` is here, but what I do know is that blockhashes are only valid _up to_ a certain slot, and that slot is given by `lastValidBlockheight`.

**Comment by jstarry** [comment]
> What happens if you set minContextSlot to lastValidBlockheight + 1?

Block height and slot are two totally different things fyi

**Comment by steveluscher** [comment]
> Block height and slot are two totally different things fyi

Oof. Of course. Sorry for the misdirection @zhe-t.

File: web3.js/test/connection.test.ts [hunk 4]
```diff
@@ -2952,6 +2952,61 @@ describe('Connection', function () {
...
+        const blockhash = await connection.getLatestBlockhashAndContext('finalized');
+
+        const isValid = await connection.isBlockhashValid(
+          blockhash.value.blockhash,
+          {
+            commitment: 'finalized',
+            minContextSlot: blockhash.context.slot+1
+          });
+        expect(isValid.value).to.be.false;
+      });
```
Comment (steveluscher) [inline web3.js/test/connection.test.ts:3001]: Do you have any advice on how to write a test that both exercises `minContextSlot` _and_ fails, @jstarry? I'm missing the plot here.

File: web3.js/src/connection.ts [hunk 5]
```diff
@@ -937,6 +937,16 @@ const BlockProductionResponseStruct = jsonRpcResultAndContext(
   }),
 );
 
+/**
+ * Configuration object for changing `isBlockhashValid` query behavior
+ */
+export type isBlockhashValidConfig = {
```
[the remark this answers is not in the document]
Comment (zhe-t) [reply web3.js/src/connection.ts]: Updated

File: web3.js/test/connection.test.ts (hunk 4 above)
Comment (jstarry) [reply web3.js/test/connection.test.ts:3001]: How about setting it to the max safe integer?

File: web3.js/src/connection.ts [hunk 6]
```diff
@@ -3359,6 +3369,45 @@ export class Connection {
...
+      [blockhash],
+      commitment || this._commitment || 'finalized',
+      'jsonParsed',
+      extra,
+    );
+
+    const unsafeRes = await this._rpcRequest('isBlockhashValid', args);
+    const res = create(unsafeRes, jsonRpcResultAndContext(boolean()));
+    if ('error' in res) {
+      throw new Error(
```
[the remark this answers is not in the document]
Comment (zhe-t) [reply web3.js/src/connection.ts:3394]: Neat. Two questions come to mind:
1. Should we check for this error code in `res` and retry a configurable amount of times before bailing? 
2. Should we return the error code be propagated in the response object of `IsBlockhashValid`?

File: web3.js/test/connection.test.ts (hunk 4 above)
Comment (zhe-t) [reply web3.js/test/connection.test.ts]: Setting to MAX_SAFE_INTEGER gives the same result 🤔 validation result is returning as true

File: web3.js/src/connection.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (jstarry) [reply web3.js/src/connection.ts:3394]: 1. yup, that's what I had in mind
2. no, it should be abstracted away

File: web3.js/test/connection.test.ts (hunk 4 above)
Comment (steveluscher) [reply web3.js/test/connection.test.ts]: I love it when writing tests pays off right away.

**Review by steveluscher — DISMISSED** [review]
Marking as request changes -> back to your queue.

File: web3.js/test/connection.test.ts (hunk 4 above)
Comment (kamlianur) [reply web3.js/test/connection.test.ts]: minContextSlot: blockhash.context.slot+1
why adding 1 (+1) after slot?

File: web3.js/src/connection.ts [hunk 7]
```diff
@@ -3359,6 +3369,35 @@ export class Connection {
...
   }
 
+  /**
+   * Check whether a blockhash is still valid or not
+   */
+  async isBlockhashValid(
+    blockhash: string,
+    config?: IsBlockhashValidConfig
+  ): Promise<RpcResponseAndContext<boolean>> {
+    const extra: Pick<IsBlockhashValidConfig, 'minContextSlot'> = {};
```
[the remark this answers is not in the document]
Comment (zhe-t) [reply web3.js/src/connection.ts:3379]: Thanks @jstarry that was it. Resolved now and all tests are passing 🎉

**Comment by zhe-t** [comment]
Updated the PR to resolve the config params not being parsed/passed correctly and now all tests are passing 🎉 thanks @jstarry @steveluscher for the guidance. Ready to review if we do not want to include the special handling for when the min context slot is not met.

**Review by steveluscher — CHANGES_REQUESTED** [review]
I pushed up a few commits to finish this up, but could you take a look at the tests @zhe-t? Write one set of tests that works in both live mode and mock mode. The key is that `mockRpcResponse` is a no-op in `LIVE` mode. Write the tests so that they would pass whether `mockRpcResponse` takes effect or not.

**Pull request closed without merging**